test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -212,34 +218,257 @@ func (s *snapshottableTestSuite) DefineTests(driver TestDriver, pattern testpatt
+		}()
+		snapshotClass, err := dc.Resource(SnapshotClassGVR).Namespace(vsc.GetNamespace()).Get(context.TODO(), vsc.GetName(), metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		snapshotClass.Object["deletionPolicy"] = "Retain"
+
+		ginkgo.By("creating a snapshot")
+		snapshot := getSnapshot(pvc.Name, pvc.Namespace, vsc.GetName())
+
+		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Create(context.TODO(), snapshot, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		defer func() {
+			framework.Logf("deleting snapshot %q/%q", snapshot.GetNamespace(), snapshot.GetName())
+			// typically this snapshot has already been deleted
+			err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Delete(context.TODO(), snapshot.GetName(), metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				framework.Failf("Error deleting snapshot %s in namespace %s. Error: %v", snapshot.GetName(), snapshot.GetNamespace(), err)
+			}
+		}()
+		err = WaitForSnapshotReady(dc, snapshot.GetNamespace(), snapshot.GetName(), framework.Poll, framework.SnapshotCreateTimeout)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("checking the snapshot")
+		// Get new copy of the snapshot
+		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Get(context.TODO(), snapshot.GetName(), metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		// Get the bound snapshotContent
+		snapshotStatus := snapshot.Object["status"].(map[string]interface{})
+		snapshotContentName := snapshotStatus["boundVolumeSnapshotContentName"].(string)
+		snapshotContent, err := dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("checking the SnapshotContent")
+		snapshotContentSpec := snapshotContent.Object["spec"].(map[string]interface{})
+		volumeSnapshotRef := snapshotContentSpec["volumeSnapshotRef"].(map[string]interface{})
+		framework.ExpectEqual(snapshotContentSpec["volumeSnapshotClassName"], vsc.GetName())
+		framework.ExpectEqual(snapshotContentSpec["deletionPolicy"], "Retain")
+		framework.ExpectEqual(volumeSnapshotRef["name"], snapshot.GetName())
+		framework.ExpectEqual(volumeSnapshotRef["namespace"], snapshot.GetNamespace())
+
+		ginkgo.By("deleting the snapshot")
+		err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Delete(context.TODO(), snapshot.GetName(), metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			framework.Failf("Error deleting snapshot %s in namespace %s. Error: %v", snapshot.GetName(), snapshot.GetNamespace(), err)
+		}
+
+		ginkgo.By("checking the SnapshotContent has not been deleted")
+		_, err = dc.Resource(SnapshotContentGVR).Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		// This check only works with the CSI hostpath driver, as the physical location of the snapshot is driver dependent.
```

Comment:
This leaks snapshots for any driver other than hostpath.

I am really unsure about having this hostpath driver specific check. If we really want to have driver-specific test logic, I think we should duplicate this test case to https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/csi_mock_volume.go, where it could potentially be added to a conformance suite in the future.  But we should keep the tests inside this storage test framework as generally applicable to all drivers as possible.


 I think we should just wait for the VolumeSnapshot

Reply:
If that's the case, we should also change the Deletion policy back to "Delete", and then delete the Content.  This way there won't be physical snapshot resources remaining.

Reply:
Makes sense, will update the test to do that.

Reply:
Updated this and added to `TestSnapshotDeleted`